python/dolfinx/fem/function.py

```diff
@@ -456,13 +467,26 @@ def _(u_ptr: int, cells: typing.Optional[np.ndarray] = None):
 
         @_interpolate.register(Expression)
         def _(expr: Expression, cells: typing.Optional[np.ndarray] = None):
-            """Interpolate Expression for the set of cells"""
-            self._cpp_object.interpolate(expr._cpp_object, cells)  # type: ignore
-
-        if cells is None:
-            mesh = self.function_space.mesh
-            map = mesh.topology.index_map(mesh.topology.dim)
-            cells = np.arange(map.size_local + map.num_ghosts, dtype=np.int32)
+            """Interpolate Expression from a given mesh onto the set of cells
+            Args:
+                expr: Expression to interpolate
+                cells: The cells to interpolate over. If `None` then all
+                    cells are interpolated over.
+            """
+            assert cell_map is not None
+            if len(cell_map) == 0:
+                # If cell map is not provided create identity map
+                assert cells is not None
+                expr_cell_map = np.arange(len(cells), dtype=np.int32)
```

Comment:
Would it be better to pass the cells in the other mesh directly? This would mean we don't need to create an identity map (you could just pass `cells` here instead). We might also not need to pass the expression mesh as well in this case

Reply:
Mesh is required due to the current way expressions are implemented (see other comment).
I thought about just supplying the cells here, but it would mean more work has to be done in Python by the user.